### ISSUE TITLE
Why choose redis as the middleware rather than ray? 

### ISSUE THREAD

**Issue by Bender1019** [seungju-mmc/Distributed_RL#1]
Thanks for this good job which enlights me a lot first. 
It seems you choose Redis as a distributed middle ware to store experience and parameters server which I haven't test yet. I am curious about why not just use ray to transfer data in the distributed setting. Is that in consideration to speed or scability problem? Is there any benchmark data? I am looking forward to your answer.

**Comment by seungju-k1m** [comment]
hello.
There are two reasons that I chose Redis.
First, I use multiple computers in order to decouple a set of actors,
learner and replay.
*I observed that if I simultaneously operate a set of actors and
learner in one computer, the overall procedure slows down.*
Perhaps it is GIL in python.
*To decouple them*,* I use multiple computers in Google Cloud
Platform*, *resulting
in the necessity of communicating between them.*
GCP supports Redis server which has high bandwidth (2GB/s) and I found that
within 64 actors, there is no bottleneck which deteriorates the quality.

Second, I don't have enough computer resources, so I have to use Compute
Engine such GCP and AWS.
Do you have 360 cpus, 256 GB RAM which are computer configuration in APE-X?
Most people don't have massive computer resources and me too.
Even though you have massive resources and operate them in only one
computer , GIL in Python makes the overall procedure slow down.

--------------------- APE-X Configuration ------------
Computer01:
256/360 cpus for sampling

Computer02:
4 high performance cpus
256GB RAMs
High performance GPU such as Tesla V100 or RTX 2070 super.
---------------------------------------------------------------------

*Notice that Two Computers not one !!*

In summary,
1. decouple a set of actors, learner and replay memory, resulting in usage
of  independent multiple computers.
2. To configure the computer resource, I have to use GCP and GCP support
for Redis which has high bandwidth.




2021년 12월 16일 (목) 오후 4:40, Yaorc ***@***.***>님이 작성:

**Comment by Bender1019** [comment]
Thanks for your reply.
That' s right. The specs of redis instance is important. I understand that you want to make use of computing resources of multiple-machine by separating actor, learner among different computers. 
My point is Ray does the same thing in a distributing setting, just like what Ray/Rllib does. So which is better? 😄
